src/ess.jl

```diff
@@ -255,37 +255,39 @@ function ess_rhat(
 )
     # compute size of matrices (each chain may be split!)
     niter = size(chains, 1) ÷ split_chains
-    nparams = size(chains, 3)
     nchains = split_chains * size(chains, 2)
     ntotal = niter * nchains
 
+    # vector used to construct output vector types
+    params = view(chains, firstindex(chains, 1), firstindex(chains, 2), :)
+
     # do not compute estimates if there is only one sample or lag
     maxlag = min(maxlag, niter - 1)
-    maxlag > 0 || return fill(missing, nparams), fill(missing, nparams)
+    maxlag > 0 || return similar(params, Missing), similar(params, Missing)
 
     # define caches for mean and variance
-    U = typeof(zero(eltype(chains)) / 1)
     T = promote_type(eltype(chains), typeof(zero(eltype(chains)) / 1))
     chain_mean = Array{T}(undef, 1, nchains)
     chain_var = Array{T}(undef, nchains)
     samples = Array{T}(undef, niter, nchains)
 
     # compute correction factor
-    correctionfactor = (niter - 1) / niter
+    correctionfactor = (niter - 1)//niter
 
     # define cache for the computation of the autocorrelation
     esscache = build_cache(method, samples, chain_var)
 
     # define output arrays
-    ess = Vector{T}(undef, nparams)
-    rhat = Vector{T}(undef, nparams)
+    ess = similar(params, T)
+    rhat = similar(ess)
+    i0 = firstindex(ess) - 1
 
     # for each parameter
     for (i, chains_slice) in enumerate(eachslice(chains; dims=3))
         # check that no values are missing
         if any(x -> x === missing, chains_slice)
-            rhat[i] = missing
-            ess[i] = missing
+            rhat[i0 + i] = missing
```

Comment:
Does that line work at all? Can `rhat` store `missing`?

Reply:
Ah, hm, not it seems this is a line currently not tested. This check should probably be moved to outside the loop.

Reply:
Ah yes, this line does work, because this conditional can only evaluate to `true` if `Missing <: eltype(chains)`, in which case `Missing <: T`.